### DISCUSSION TITLE
PSA: Prisma Client generation is really wonky on Vercel

### DISCUSSION THREAD

**Discussion by leo-petrucci** [prisma/prisma#5140] [Q&A] [answered]
I've been hitting my head on the wall for days trying to figure out why my API didn't work when deploying on Vercel despite working locally, but I've figured out why and I think it should be mentioned in the related repo for the example, but please let me know if I'm just a moron and this should've been obvious.

### Deploying prisma to Vercel breaks

So I've started my project from the [official example found here](https://github.com/prisma/prisma-examples/tree/latest/deployment-platforms/vercel-graphql). This project works perfectly when deployed to Vercel directly because it forces Vercel to download, build and generate the prisma client.

Unfortunately, if you were to upload your project to Github and set up auto-deploys on commit, you will soon find a bunch of errors related to `@prisma/client` missing exports in your build outputs. For me these were missing types for my data and even native exports like `empty` and `sql`.

If you were to try to access your API you would most likely be able to see something like `2021-01-17T16:41:33.659Z	undefined	ERROR	Error: You most likely forgot to initialize the Prisma Client. Please run `prisma generate` and try to run it again.` in your function logs.

### Why this happens

Vercel, like many other serverless hosts, uses a `package-lock`-based cache. What this means in practice is:

1. Change is made to the Github
2. Vercel sees the changes
3. Checks if anything has changed in `package-lock`
4. If it's the same as the previous commit, it downloads a cache of `node_modules`

**Unfortunately**, `node_modules` is where the prisma client (which contains all of our types) is kept. And even more unfortunately, changes to our API which would be reflected in our types when running `npm run generate` does not make changes to `package-lock`.

In short this means that even though our app may have gone through significant changes, if `package-lock` isn't deleted manually, our types will not be generated, which in turn will most likely break our API.

### Solution

As previously said, for now I think the only solution is to manually delete `package-lock`. But I don't think `.prisma/client` should be created inside `node_modules`.

Other frameworks which need to generate their own files in a project to function (eg Next.js) usually create their folders inside the root of the project (eg `/.next`), so maybe that could be a better place to create it?

Either way, I hope this helps someone so they don't have to go through what I did.

**Comment by ryands17** [comment]
Hey @creativiii 👋 

Does adding a `postinstall` script with the command `prisma generate` to your `package.json` work? If not, you can always add this as a `prebuild` step if Vercel has a build step.

**Reply by leo-petrucci** [reply]
Postinstall hook tells me it couldn't find my schema, despite it being in the default location.

Haven't tried to put it in prebuild but I'll give it a go. Although I don't think it's going to make much difference as I'm running generate on build and that still doesn't seem to work.

**Reply by ryands17** [reply]
Generate should work on `build`. What error are you getting on `build`?

**Reply by leo-petrucci** [reply]
It should, but it doesn't seem to, not until I overrwrite the cached `node_modules`. Here's some of the stuff I get:

```

NowBuildError: api/_lib/query.ts(1,10): error TS2305: Module '"../../node_modules/@prisma/client"' has no exported member 'empty'.
--
00:05:29.011 | api/_lib/query.ts(1,17): error TS2305: Module '"../../node_modules/@prisma/client"' has no exported member 'sql'.
00:05:29.011 | at createTSError (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:99067:16)
00:05:29.011 | at reportTSError (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:99073:23)
00:05:29.011 | at getOutputTypeCheck (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:99160:17)
00:05:29.011 | at compile (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:99238:40)
00:05:29.012 | at compileTypeScript (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:98515:31)
00:05:29.012 | at Job.readFile (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:98541:30)
00:05:29.012 | at Job.emitDependency (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:48057:33)
00:05:29.012 | at /vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:48095:32
00:05:29.012 | at Array.map (<anonymous>)
00:05:29.012 | at Job.emitDependency (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:48075:26) {
```

```

NowBuildError: api/_lib/utils.ts(4,10): error TS2305: Module '"../../node_modules/@prisma/client"' has no exported member 'Prisma'.
--
00:05:29.689 | api/_lib/utils.ts(4,18): error TS2305: Module '"../../node_modules/@prisma/client"' has no exported member 'User'.
00:05:29.689 | at createTSError (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:99067:16)
00:05:29.689 | at reportTSError (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:99073:23)
00:05:29.689 | at getOutputTypeCheck (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:99160:17)
00:05:29.689 | at compile (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:99238:40)
00:05:29.689 | at compileTypeScript (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:98515:31)
00:05:29.690 | at Job.readFile (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:98541:30)
00:05:29.690 | at Job.emitDependency (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:48057:33)
00:05:29.690 | at /vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:48095:32
00:05:29.690 | at Array.map (<anonymous>)
00:05:29.690 | at Job.emitDependency (/vercel/f3316494f3c73b11/.build-utils/.builder/node_modules/@vercel/node/dist/index.js:48075:26) {
```

I also used to get errors from my defined models `t.model.something()` because they didn't exist. I thought this was my mistake so I changed them all to `t.field('something', { type: 'String' })` but that would still give me errors later on because the types weren't generated.

Note that the build succeeds despite the above errors, but the actual API will respond with a 502.

It's also very possible Vercel has a build cache as well, which would mean nothing is getting regenerated because the `package-lock` file hasn't changed 😂

**Reply by ryands17** [reply]
@creativiii 👋 
I have created an issue here https://github.com/prisma/prisma/issues/5175 and the team will be adding a workaround for this :)

**Reply by 2color** [reply, marked as answer]
As discussed in #5175, adding the `postinstall` hook solved the problem after all.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
